src/main/kotlin/com/haroldadmin/cnradapter/NetworkResponse.kt

```diff
@@ -1,61 +1,140 @@
+ * otherwise.
+ *
+ * If the network request was successful and Retrofit successfully deserialized the body, the [NetworkResponse]
+ * is [NetworkResponse.Success]. If you do not expect a successful response to contain a body, you must specify
+ * [S] as [Unit] or use [CompletableResponse].
+ *
+ * If the network request failed due to:
+ * - Non-2xx response from the server, the [NetworkResponse] is [NetworkResponse.ServerError] containing the
+ * deserialized body of the response ([E])
+ * - Internet connectivity problems, the [NetworkResponse] is [NetworkResponse.NetworkError]
+ * - Any other problems (e.g. Serialization errors), the [NetworkResponse] is [NetworkResponse.UnknownError].
+ */
+public sealed interface NetworkResponse<S, E> {
     /**
-     * A request that resulted in a response with a 2xx status code that has a body.
+     * The result of a successful network request.
+     *
+     * If you expect your server response to not contain a body, set the success body type ([S]) to [Unit].
+     * If you expect your server response to sometimes not contain a body (e.g. for response code 204), set
+     * [S] to [Unit] and deserialize the raw [response] manually when needed.
+     *
+     * @param body The parsed body of the successful response.
+     * @param response The original [Response] from Retrofit
      */
-    data class Success<T : Any>(
-        val body: T,
-        val headers: Headers? = null,
-        val code: Int
-    ) : NetworkResponse<T, Nothing>()
+    public data class Success<S, E>(
+        public val body: S,
+        public val response: Response<*>
+    ) : NetworkResponse<S, E> {
+        /**
+         * The status code returned by the server.
+         *
+         * Alias for [Response.code] of the original response
+         */
+        public val code: Int
+            get() = response.code()
+
+        /**
+         * The headers returned by the server.
+         *
+         * Alias for [Response.headers] of the original response
+         */
+        public val headers: Headers
+            get() = response.headers()
+    }
 
     /**
-     * Describe an error without a specific type.
-     * Makes it easier to deal with the case where you just want to know that an error occurred,
-     * without knowing the type
+     * The result of a failed network request.
      *
-     * @example
-     * val response = someNetworkAction()
-     *
-     * when (response) {
-     *    is NetworkResponse.Success -> // Action Succeeded do something with body
-     *
-     *    is NetworkResponse.Error -> // Action failed do something with error
-     * }
+     * A failed network request can either be due to a non-2xx response code and contain an error
+     * body ([ServerError]), or due to a connectivity error ([NetworkError]), or due to an unknown
+     * error ([UnknownError]).
      */
-    interface Error {
-        val error: Throwable
+    public sealed interface Error<S, E> : NetworkResponse<S, E> {
```

Comment:
Is it necessary to carry the `S` into `Error`? `ServerError` contains a `Response`, but does not enforce its type. It's not a big deal, but I am curious if there is a way to simplify the `Error` interface.

Reply:
Unfortunately, yes. Without the generic parameter `S` in `Error` you have to typecast quite a bit in tests.